### ISSUE TITLE
Walkthrough

### ISSUE THREAD

**Issue by devDesu** [wapiflapi/exrs#1]
Some guide for solving reverse exercices could be very helpful.

**Comment by joshfaust** [comment]
Yes, has there been any walktroughs? I would live a little guidence as I am new to RE.

**Comment by M0S4YSH3LL0** [comment]
I can provide walkthrough for r1-r4, haven't solved the others yet.

**Comment by bernardyip** [comment]
I have completed r1 to r8 and would love a hint or some push for r9 if possible!

**Comment by devDesu** [comment]
> I have completed r1 to r8 and would love a hint or some push for r9 if possible!

So basicly r9 consists of two tasks:
Solve series (restore k buffer)
x{0} = 0x1337
x{i+1} = x{i} * (k{i} & 0xFF) + 0x31337
x{final} = 0xFD0970E7
final = strlen(input)

After that restore original input knowing that random() returns same values on each run

Right now I'm stuck solving this equation/series

Here's my Python code to solve this, but it's not fast enough
--------------------
```
def find_dividers(x):
    return ((i,x/i) for i in range(1, 0x100) if x%i == 0)
    
def find_solutions(val, solutions, depth=1):
    start_val = 0x1337
    if depth > 20:
        return -1
    if val == start_val:
        print('Good solution: ', solutions)
        return 1
    elif val < start_val:
        return -1
    for divider, value in find_dividers(val):
        find_solutions(value-0x31337, "%s.%d"  % (solutions, divider), depth+1)

end_val = 0xFD0970E7
find_solutions(end_val - 0x31337, "")
```

**Comment by bernardyip** [comment]
@devDesu

I found several 'passwords' that work through bruteforcing. One of which is: 6`SHQe

However i feel that this is not the correct way to solve this exercise and there has to be a way, via some weakness, to break the code without dynamic analysis (which we required to get the randoms). This feeling is felt as the previous 8 exercise does not require dynamic analysis of any kind.

Some possibilities that i thought of:
- making strdup fail in order to get return false immediately
- sending an empty string, which skips the xoring issue (meaning we dont have to care about the randoms at all!) But we have an issue controlling k
- I realise the formula of k is actually 0x31337 * (1 + a + ab + abc + abcd...) where abcd.... are the xor-ed results of the input. Not sure if thats a good hint.
- the multiplications done above take into consideration the sign bits which would mean sometimes the value can be a negative

Would love to see if anything else was missed out or some new hint which can help with breaking of the code

**Comment by devDesu** [comment]
@bernardyip 
I don't think that we need to try to exploit something
Looks like this task show us "random prediction" when you forget to give different seed to generator
But I don't really like bruteforcing part and don't know if we miss something

If you are interested in simillar task or just want to talk about RE you can write me to temporary email
gitgitgit.220011 at gmail dot com

**Comment by wapiflapi** [comment]
I really appreciate y'all's conversation :) I made these challenges a LONG time ago and I don't remember exactly if there was a better / expected solution for `r9`. I think the the intent was indeed to introduce random prediction but I'm not sure the challenge was designed well enough that it's easily possible to recover the original key. Constraint solving with reduced character sets might be a possibility. (I haven't tested it.) Anyway, if the password works I guess it counts as a success ;-)

Closing this issue since I don't think I'm willing to put in the work to "fix" the challenge, if any one wants to submit a different version of the challenge with less ambiguity PRs are always welcome!

**Comment by jeffli678** [comment]
For r9, first use z3 to find a string that satisfy the equation. And then undo the xor, which gives you the original string. I have not do it yet but it should work.

**Comment by eranzim** [comment]
Regarding r9:
The "correct" way to go is probably smt z3, as @jeffli678 suggested. I managed to brute-force it with python instead (probably simpler to people not familiar with z3). It's important to notice that when the character from the xored input is multiplied with the "accumulator" (the loop's index / k), it is first sign-extended to 32-bit, so in fact we should treat the input characters as -128..127 rather than 0..255. Brute-forcing input strings of size 1-3 didn't find a result, but here is my code to solve it with strings of length 4 (where several solutions are found):

```
for i in range(-128,128):
    for j in range(-128,128):
        for k in range(-128,128):
            for m in range(-128,128):
                x = ctypes.c_int(m*(k*(j*((i*4919)+201527)+201527)+201527)+201527).value
                if x == -49712921:
                    print(m, k, j, i)
```
This code should start producing results within a few minutes, for example: -2 9 -7 -115.
Then I found out what the input of length 4 is xored with; I gave an input of "0000", saw it becomes 8F 2F F2 AB, and since ord('0')==48, the first input character is then 48^0x8f^0xfe (0xfe being -2) which is 'A'. And so on, feel free to contact me for a full solution :)
Do note - I found 5 solutions of length 4, none of which contain only printable ASCII chars. It is possible to brute force only those solutions (rather than taking the whole range(-128,128)), which probably needs length 5 or more, or indeed using z3, but I stopped at "Password OK".
